### PR TITLE
fix rem2pi for non-finite arguments

### PR DESCRIPTION
**Pull request by oscardssmith**
followup to https://github.com/JuliaLang/julia/pull/36420 that fixes `Inf` and `-Inf` as well (and I believe at slightly lower cost)

### CODE REVIEW

**Comment by oscardssmith** [comment]
We also should really make better implimentations of `rem2pi` for `Float32` and `Float16`. Our current methods just use the `Float32` impl, which is pretty slow.

File: test/numbers.jl [hunk 1]
```diff
@@ -2676,10 +2676,11 @@ end
 end
 
 @testset "PR #36420 $T" for T in (Float16, Float32, Float64)
-    @test rem2pi(T(NaN), RoundToZero)  === T(NaN)
-    @test rem2pi(T(NaN), RoundNearest) === T(NaN)
-    @test rem2pi(T(NaN), RoundDown)    === T(NaN)
-    @test rem2pi(T(NaN), RoundUp)      === T(NaN)
+    for r in (RoundToZero, RoundNearest, RoundDown, RoundUp)
+        for x in (Inf, -Inf, NaN, -NaN)
+            @test isnan(rem2pi(T(x), r))
```
Comment (giordano) [inline test/numbers.jl:2681]: This doesn't check the type of NaN though, right?

File: base/math.jl [hunk 2]
```diff
@@ -1239,7 +1239,7 @@ julia> rem2pi(7pi/4, RoundDown)
 """
 function rem2pi end
 function rem2pi(x::Float64, ::RoundingMode{:Nearest})
-    isnan(x) && return NaN
+    (x-x !== 0.0) && return NaN
```
Comment (giordano) [inline base/math.jl:1242]:
```suggestion
    !isfinite(x) && return NaN
```

Comment (giordano) [reply base/math.jl:1242]: At the moment `isfinite(::AbstractFloat)` is https://github.com/JuliaLang/julia/blob/783a6aa4d3e5ad371a5cec5d7510f46d423042e3/base/float.jl#L497 but perhaps it should be `x - x !== zero(x)`, if that's more efficient?  Or even `iszero(x - x)`, and `iszero` should be `x === zero(x)`

**Comment by oscardssmith** [comment]
ready to merge?

**Comment by giordano** [comment]
Use `isfinite` everywhere else?

**Comment by oscardssmith** [comment]
good point.

**Pull request merged**